pkg/cdi/cache.go

```diff
@@ -277,3 +339,113 @@ func (c *Cache) GetSpecDirectories() []string {
 	copy(dirs, c.specDirs)
 	return dirs
 }
+
+// Our fsnotify helper wrapper.
+type watch struct {
+	w       *fsnotify.Watcher
+	watched map[string]bool
+}
+
+// Start watching a set of directories for relevant changes.
+func (w *watch) start(dirs []string, m *sync.Mutex, refresh func() error) error {
+	var (
+		dir string
+		err error
+	)
+
+	w.watched = make(map[string]bool)
+
+	w.w, err = fsnotify.NewWatcher()
+	if err != nil {
+		return errors.Wrap(err, "failed to create Spec dir watch")
+	}
+
+	for _, dir = range dirs {
+		w.watched[dir] = false
+	}
+
+	w.update()
+	go w.watch(m, refresh)
+	refresh()
+
+	return nil
+}
+
+// Stop watching directories.
+func (w *watch) stop() {
+	if w.w == nil {
+		return
+	}
+
+	w.w.Close()
+	w.watched = nil
+}
+
+// Update watch with pending/missing or removed directories.
+func (w *watch) update(removed ...string) bool {
+	var (
+		dir    string
+		done   bool
+		err    error
+		update bool
+	)
+
+	for dir, done = range w.watched {
+		if done {
+			continue
+		}
+		err = w.w.Add(dir)
+		switch {
+		case err == nil:
+			w.watched[dir] = true
+			update = true
+		case !os.IsNotExist(err):
+			fallthrough
+		default:
+			w.watched[dir] = false
+		}
+	}
+
+	for _, dir = range removed {
+		w.watched[dir] = false
+		update = true
+	}
+
+	return update
+}
+
+// Watch Spec directory changes, triggering a refresh if necessary.
+func (w *watch) watch(m *sync.Mutex, refresh func() error) {
+	watch := w.w
+	for {
+		select {
+		case event, ok := <-watch.Events:
+			if !ok {
+				return
+			}
+
+			if (event.Op & (fsnotify.Rename | fsnotify.Remove | fsnotify.Write)) == 0 {
+				continue
+			}
+			if event.Op == fsnotify.Write {
+				if ext := filepath.Ext(event.Name); ext != ".json" && ext != ".yaml" {
+					continue
+				}
+			}
+
+			m.Lock()
+			var removed []string
```

Comment:
There is no need to construct this slice since `update` accepts a variadic list.

Reply:
That was done that way to set up the arguments in the if-else then call `update()` on the common path... which is not that helpful. Updated pass args without a slice and the ...-operator.